### ISSUE TITLE
Facing a problem with the orientation of the whycon tag.

### ISSUE THREAD

**Issue by damounayman** [jiriUlr/whycon-ros#2]
Hello, I am very interested in your Fiducial marker. I am trying to get the position and orientation of an object using Fiducial markers. For the position, it's perfect, but the orientation, it's wrong, there is an error of 20 to 30 deg in the Euler angle.

Can you give me a hint on where the problem is?

I use a Realsense D435i camera and the calibration settings are fed to the ROS node.

**Comment by jiriUlr** [comment]
Hi, it's hard to tell from such a brief description. Do you use WhyCon or WhyCode marker? WhyCon supports mainly position estimation. Can you provide the sample data and configuration you are using?

**Comment by damounayman** [comment]
HI, thank you for replying.
I'm using whycode marker. 
for seek of simplicity, I'm using:
 <img src="https://user-images.githubusercontent.com/67443398/182833415-f1d2337d-0a1b-4a92-9096-95a78ac03db4.png" width="100" height="100">
For the generation, I used  your  [git repo](https://github.com/jiriUlr/whycode-gen)
```bash
./whycode_gen -t 2
```
For the configuration of the launch file and ros param, I choose the following config:

- circleDiameter : 0.009 m
- numMarkers : 4 
- minSize : 30
- fieldLength : 1.0
- idBits : 2
- idSamples: 4
- hammingDist : 1
- maxPatterns : 1
- areaRatioTolerance: 40.0
- initCircularityTolerance : 100.0
- finalCircularityTolerance : 2.0
- centerDistToleranceRatio: 10
- centerDistToleranceAbs: 5.0

For the Realsense D435i  camera parameters provided by the camera_info topic, I have for 1280x720:
distortion matrix D: [0.0, 0.0, 0.0, 0.0, 0.0]
intrinsic matrix K: [908.1417846679688, 0.0, 642.3192749023438, 0.0, 908.0018310546875, 369.9261169433594, 0.0, 0.0, 1.0]

When I run the code with this config, I get these results:
![demo_whycode](https://user-images.githubusercontent.com/67443398/182840359-a37db092-f6eb-46df-8449-61716eb9be7e.gif)

```
header: 
  seq: 29707
  stamp: 
    secs: 1659613973
    nsecs: 983729336
  frame_id: "camera_color_optical_frame"
markers: 
  - 
    id: 1
    size: 1077
    u: 640.9136352539062
    v: 310.3704833984375
    angle: 2.3525354862213135
    position: 
      position: 
        x: 0.21919476985931396
        y: 0.0003418896230868995
        z: 0.014386740513145924
      orientation: 
        x: 0.9554481506347656
        y: 0.0019961483776569366
        z: -0.05576407164335251
        w: 0.28983646631240845
    rotation: 
      x: 2.5509862899780273
      y: 0.10792586952447891
      z: -0.028681274503469467
```

**Comment by jiriUlr** [comment]
Hi,
I'd suggest leaving the default value for `id_samples` parameter e.g. 360 or something like that. I don't understand where you got the parameter `maxPatterns`. When I freshly clone and run the node, I don't see it by `rosparam list`. Is the quaternion of the resulting 6DOF pose also wrong or is it only the Euler angles representation?

**Comment by damounayman** [comment]
Hi,
I clone the repo and I change only the diameter to 0.009 m.
After running the node, I get these results. The position is perfect, but not the orientation.
![position_whycon](https://user-images.githubusercontent.com/67443398/182937498-126fccc7-eb8a-4326-b765-cdc31ceaa3bd.png)

I experimented manually by changing the orientation of the marker and at this orientation, the code returns that the tag has the same orientation as the camera. 

![whycon](https://user-images.githubusercontent.com/67443398/182938617-797160ad-757b-456f-aa9a-23036b9994ec.gif)

 Is this oscillation normal?

```cpp
void CTransformation::calcOrientation(STrackedObject &obj)
{
    calcQuaternion(obj);
    calcEulerFromQuat(obj);
}
```
I verify the CalcEulerFromQuat function with eigen library and I get the same result. 
So I think the problem is with calcQuaternion or its input.

**Comment by uzgit** [comment]
Hi Jiri, Ayman,

I am subscribed to the WhyCode stuff and have done some work in the past to
try to mitigate the issue of oscillating orientations, so this issue was
sent to me as well.

If you want, you can check my repo here:
https://github.com/uzgit/whycon-ros/tree/ellipse_sampling

It is a fork of Jiri's repo that I edited to do the orientation decision
differently. In my experiments it seems to have less oscillation. (I have
tested 8-bit markers.)

Best,
Joshua

**Comment by damounayman** [comment]
Hi Jiri, uzgit
 tanks uzgit for sharing your work.
So just by using a marker with 3 bits, I get a better orientation than with 2 bits.
![save_marker](https://user-images.githubusercontent.com/67443398/183031351-82e88dbc-24e1-4477-b1e0-f61ad377f13f.png)
I tested your uzgit version, but I still have that oscillation in orientation.
![save_marker](https://user-images.githubusercontent.com/67443398/183034835-663c632f-85e4-463a-aef2-045bf538c8f1.gif)

Can you tell me where I can act to improve the stability of the orientation?

**Comment by uzgit** [comment]
You can try using a higher number of bits - I’d suggest 8

**Comment by uzgit** [comment]
And be sure to use the ellipse_sampling branch

**Comment by jiriUlr** [comment]
> I experimented manually by changing the orientation of the marker and at this orientation, the code returns that the tag has the same orientation as the camera.

It should be fixed. It was only a matter of one more simple quaternion.

> Is this oscillation normal?

The case you are showing is a bit difficult for orientation estimation as the marker's plane is parallel to the image plane and the markers are close to the principal point. Thus, the currently applied approach has a hard time distinguishing the correct orientation.
Therefore, you might try Joshua's approach, which uses a different method to select the correct pose.
A bigger marker or more bits may also help.

I would appreciate it if you could provide a small rosbag of the camera output so that I could examine the issue closely.

> Can you tell me where I can act to improve the stability of the orientation?

The ambiguity in orientation is resolved here.
https://github.com/jiriUlr/whycon-ros/blob/master/src/whycon/CCircleDetect.cpp#L602

The oscillation in the orientation could be resolved by an adaptive temporal filter of the marker's pose. Considering that it oscillates between two orientation candidates, one could measure the rotation difference of consecutive estimates and select the other pose from the ambiguity pair when the error would be above a certain threshold.
Also, the slight variation of correctly estimated poses could be smoothed by adaptive weighted averaging of the whole pose.



If you want, you can reach me at ulricji1@fel.cvut.cz, and we can schedule a call to go through the possibilities, settings and setup of WhyCode.

**Comment by damounayman** [comment]
Hi [jiriUlr](https://github.com/jiriUlr),[uzgit](https://github.com/uzgit) 
Thank you for your answers. I will follow your indications to improve the orientation. However, I have the constraint of not exceeding 2 cm regarding the diameter of the marker. I will be happy to discuss this with you in person. I will send you the rosbag on your email.

> You can try using a higher number of bits - I’d suggest 8
> And be sure to use the ellipse_sampling branch

Yes, I tested your approach [uzgit](https://github.com/uzgit). It still wobbles, I think the problem is with the size of the marker I chose. I'll dig deeper into your approach to understand better.
